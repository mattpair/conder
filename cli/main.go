package main

import (
	// "./cmd"
	"fmt"
	"io/ioutil"
	"net/http"
)

func main() {
	out, err := http.Get("https://localhost:7344")
	if err != nil {
		fmt.Println("error", err)
		return
	}
	defer out.Body.Close()
	if out != nil {
		fmt.Println(out)
	}
	body, err := ioutil.ReadAll(out.Body)

	if err != nil {
		fmt.Println("error", err)
	}
	if body != nil {
		fmt.Println(out)
	}

	// cmd.Execute()
}
