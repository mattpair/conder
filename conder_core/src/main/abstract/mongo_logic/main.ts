import { ActionSequence, calculate_lock_requirements, LockRequirements } from './lock_calculation';
import { MongoNodeSet } from '../globals/mongo';
import { TargetNodeSet, NodeSet } from "../IR";
import { ScopeMap } from '../../data_structures/scope_map';
import { MONGO_ACTION_SUMMARIZER } from './action_summarizer';

type LockCalculator = (input: Map<string, TargetNodeSet<MongoNodeSet>[]>) =>  Map<string, LockRequirements[string]>

export const MONGO_LOCK_CALCULATOR: LockCalculator = input => {
    
    const actions: Record<string, ActionSequence> = {}
    input.forEach((v, k) => {
        actions[k] = MONGO_ACTION_SUMMARIZER(v)
    })
    const locks = calculate_lock_requirements(actions)
    
    return new Map(Object.entries(locks))
}

