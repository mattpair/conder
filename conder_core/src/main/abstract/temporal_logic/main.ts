type Actions = {
    get: {id: string},
    mutation: {id: string, using: string[]},
}
type ActionKind = keyof Actions
type AnyAction = {
    [K in keyof Actions]: {kind: K} & Actions[K]
}[keyof Actions]


export type ActionSequence = AnyAction[]
export type LockRequirements = Record<string, {global: string, kind: "r" | "w"}[]>

export function calculate_lock_requirements(sequences: Record<string, ActionSequence>): LockRequirements {
    const actionAgainstData = new Map<string, Set<ActionKind>>()
    const lockReqs: LockRequirements = {}

    Object.values(sequences).forEach(seq => {
        seq.forEach(action => {
            const ops = actionAgainstData.get(action.id)
            if (ops === undefined) {
                actionAgainstData.set(action.id, new Set<ActionKind>([action.kind]))
            } else {
                ops.add(action.kind)
                actionAgainstData.set(action.id, ops)
            }
        })
    })
    
    Object.keys(sequences).forEach(func => {
        lockReqs[func] = []
        const previouslyGot = new Set<string>()
        sequences[func].forEach(action => {
            
            switch (action.kind) {
                case "get":
                    if (previouslyGot.has(action.id)) {
                        if (actionAgainstData.get(action.id).has("mutation")) {
                            lockReqs[func].push({global: action.id, kind: "r"})
                        }
                    } else {
                        previouslyGot.add(action.id)
                    }
                case "mutation":
                    break
                default: 
                    const n: never = action
            }
        })
    })

    return lockReqs
}