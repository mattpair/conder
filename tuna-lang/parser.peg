globals := {value={varDecl | func} {ws+ | '$'}}*

mutable := 'let'
constant := 'const'
varDecl := ws* mutability={ constant | mutable } ws+ name=name ws* equals ws* value=expression
name := name='[a-zA-Z]+\w*'
equals := '='
ws := '\s'
obj := '\{' ws* fields=fields ws* '\}'
str := '\'' value='[\w \t]*' '\''
bool := value={'true' | 'false'}
num := value='-?\d+(\.\d+)?'
literal := obj | str | bool | num
newLineOrComma := '\n|,'
field := name=name ws* ':' value=literal newLineOrComma?
fields := value=field*
space := ' '

executable := ws* value={value={ret | varDecl | assignment | expression } ws+ }*
expression := !ret root={literal | name   } methods={method}*
method := method={parameterIndex | literalIndex}

parameterIndex := '\[' space* value={expression} space* '\]'
literalIndex := '\.' value={name}
objectIndex := obj=expression index={parameterIndex | literalIndex}
ret := 'return(?=\s)' value={expression}?
func := ws* 'public' space+ 'function' space+ name=name space* args=args ws* '\{' body=executable '\}'
args := '\(' ws* leadingArgs={name=name newLineOrComma ws*}* ws* lastArg={name}? ws* '\)'
assignment := target=expression space* equals space* value=expression