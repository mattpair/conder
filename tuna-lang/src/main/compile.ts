import { ParseResult, globals_$0_$0, ASTKinds } from "./parser";


export function compile(p: ParseResult): globals_$0_$0[] {
    if (p.err) {
        throw Error(`Failure parsing: line ${p.err.pos.line} col ${p.err.pos.offset}`)
    }
    return p.ast.map(g => {
        switch (g.value.kind) {
            case ASTKinds.varDecl: 
                if (g.value.mutability !== "const") {
                    throw Error(`Global variable ${g.value.name.name} must be const`)
                }
            case ASTKinds.func: 

                return g.value

            default: 
                const ne: never = g.value
        }
    })

}