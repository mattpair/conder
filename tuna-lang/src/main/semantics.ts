import { ParseResult, executable, ASTKinds, func, expression, literal } from "./parser";
import {AnyNode, PickNode, FunctionDescription} from "conder_core"

type GlobalObject = {kind: "glob", name: string}

export type Manifest = {
    globals: Map<string, GlobalObject>
    funcs: Map<string, FunctionDescription>
}

type ScopeMapEntry= "func" | "global" | {kind: "const" | "mut", index: number}
class ScopeMap extends Map<string, ScopeMapEntry>  {
    nextVar: number = 0
    
    public get nextVariableIndex(): number {
        return this.nextVar
    }

    set(k: string, v: ScopeMapEntry): this {
        if (v !== "global" && v !== "func") {
            this.nextVar++
        }
        super.set(k, v)
        return this
    }


}


function excluding<EX extends AnyNode["kind"]>(node: AnyNode, ...not: EX[]): Exclude<AnyNode, {kind: EX}> {
    //@ts-ignore
    if (not.includes(node.kind)) {
        throw Error(`Invalid expression ${node.kind}`)
    }
    //@ts-ignore
    return node
}
function only<IN extends AnyNode["kind"]>(node: AnyNode, ...only: IN[]): Extract<AnyNode, {kind: IN}> {
    //@ts-ignore
    if (only.includes(node.kind)) {
        //@ts-ignore
        return node
    }

    throw Error(`Invalid expression ${node.kind}`)
}

function cbOnly<IN extends AnyNode["kind"]>(node: AnyNode, f: (i: PickNode<IN>) => void, ...only: IN[]) {
    //@ts-ignore
    if (only.includes(node.kind)) {
        //@ts-ignore
        f(node)
    }
}


function literal_to_node(lit: literal, scope: ScopeMap): AnyNode {
    switch (lit.kind) {
        case ASTKinds.bool:
            return {kind: "Bool", value: lit.value === "true"}
        case ASTKinds.obj:
            
            return {
                kind: "Object", 
                fields: lit.fields.value.map(f => {
                
                    return {
                        kind: "SetField", 
                        field_name: [{kind: "String", value: f.name.name}], 
                        value: excluding(literal_to_node(f.value, scope),  "Return", "SetField", "If", "Save", "Update")
                    }
                })
            }
        case ASTKinds.str:
            return {
                kind: "String",
                value: lit.value
            }
        case ASTKinds.num:    
            return {
                kind: "Int",
                value: parseFloat(lit.value)
            }
    }
}

function method_to_node(target: PickNode<"Saved" | "GlobalObject">, methods: expression["methods"], scope: ScopeMap): AnyNode {
    if (methods.length === 0) {
        return target
    }
    const fields: PickNode<"GetField">["field_name"] = []
    methods.forEach(m => {
        switch(m.method.kind) {
            case ASTKinds.literalIndex: 
                fields.push({kind: "String", value: m.method.value.name})
        
                break
            case ASTKinds.parameterIndex:
                const e = expression_to_node(m.method.value, scope)
                fields.push(only(e, "String", "Saved"))
                break

            default: 
                const n: never = m.method
        }
    })
    
    return {
        kind: "GetField",
        target,
        field_name: fields
    }
}

type Target = {root: PickNode<"Update">["target"], field: PickNode<"SetField">["field_name"]}
function expression_to_update_target(exp: expression, scope: ScopeMap): Target {

    switch (exp.root.kind) {
        case ASTKinds.name:
            const name = scope.get(exp.root.name)
            if (name === undefined) {
                throw Error(`Unrecognized name ${exp.root.name}`)
            }
            if (name === "func") {
                throw Error(`Invalid function reference ${exp.root.name}`)
            }
            if (name !== "global" && name.kind === "const") {
                throw Error(`Attempting to overwrite constant variable ${exp.root.name}`)
            }
            
            const root: Target["root"] = name === "global" ? 
                {kind: "GlobalObject", name: exp.root.name} 
                : {kind: "Saved", index: name.index}
            
            const field: Target["field"] = exp.methods.map(m => {
                switch (m.method.kind) {
                    case ASTKinds.literalIndex:
                        return {
                            kind: "String",
                            value: m.method.value.name
                        }
                        
                    case ASTKinds.parameterIndex:
                        return only(expression_to_node(m.method.value, scope), "String", "Saved")
                    default: 
                        const n: never = m.method
              }
            })
            return {root, field}


        default: throw Error(`Invalid assignment to ${exp.root.kind}`)
    }
}

function expression_to_node(exp: expression, scope: ScopeMap): AnyNode {
    switch(exp.root.kind) {
        case ASTKinds.bool:
        case ASTKinds.str:
        case ASTKinds.num:
        case ASTKinds.obj:
            if (exp.methods.length > 0) {
                throw Error(`Unexpected method on ${exp.root.kind} literal`)
            }
            return only(literal_to_node(exp.root, scope), "Bool", "Int", "String", "Object")
            
        
        case ASTKinds.name:
            const name = scope.get(exp.root.name)
            if (name === undefined) {
                throw Error(`Unrecognized name ${exp.root.name}`)
            }
            if (name === "func") {
                throw Error(`Invalid function reference ${exp.root.name}`)
            } else if (name === "global") {
                return method_to_node({kind: "GlobalObject", name: exp.root.name}, exp.methods, scope)
            } else {
                return method_to_node({kind: "Saved", index: name.index}, exp.methods, scope)
            }
        default: 
            const n: never = exp.root
    }
    
}

function to_computation(ex: executable, scope: ScopeMap): FunctionDescription["computation"] {
    const ret: FunctionDescription["computation"] = []
    ex.value.forEach(e => {
        
        switch (e.value.kind) {
            case ASTKinds.ret:
                ret.push({
                    kind: "Return", 
                    value: e.value.value !== null ? excluding(expression_to_node(e.value.value.exp, scope), "Update", "SetField", "Return", "If", "Save") : undefined
                    }
                )
                break
            

            case ASTKinds.assignment: {
                const target: Target = expression_to_update_target(e.value.target, scope)
                const value: PickNode<"Update">["operation"] = excluding(expression_to_node(e.value.value, scope), "Update", "Return", "If", "Save", "SetField")
                
                ret.push({
                    kind: "Update",
                    target: target.root,
                    operation: target.field.length > 0 ? {kind: "SetField", field_name: target.field, value} : value
                })
                break
            }
            case ASTKinds.expression: {
                cbOnly(expression_to_node(e.value, scope), ret.push, "Return", "If", "Save", "Update")
                
                break
            }

            case ASTKinds.varDecl: 
                const value = excluding(expression_to_node(e.value.value, scope), "Return", "If", "Update", "Save", "SetField")
                const index = scope.nextVariableIndex
                if (scope.has(e.value.name.name)) {
                    throw Error(`The symbol ${e.value.name.name} is already in use`)
                }
                scope.set(e.value.name.name, {kind: e.value.mutability === "const" ? "const" : "mut", index})
                ret.push({
                    kind: "Save",
                    value,
                    index
                })
                break
            default: 
                const n: never = e.value
        }
    })
    return ret
}

function to_descr(f: func, scope: ScopeMap): FunctionDescription {
    try {
        const argList: string[] = []
        if (f.args.leadingArgs.length > 0) {
            argList.push(...f.args.leadingArgs.map(a => a.name.name))
        }
        if (f.args.lastArg) {
            argList.push(f.args.lastArg.name)
        }
        const input: FunctionDescription["input"] = []
        argList.forEach((a, i) => {
            if (scope.has(a)) {
                throw Error(`Arg ${a} is using a name already in use`)
            }
            scope.set(a, {kind: "mut", index: i})
            input.push({kind: "Any", data: undefined})
        })
        
        return {
            input,
            computation: to_computation(f.body, scope)
        }
    } catch(e) {
        throw Error(`In function ${f.name.name}: \n\t${e.message}`)
    }
    
}

export function semantify(p: ParseResult): Manifest {
    if (p.err) {
        throw Error(`Failure parsing: line ${p.err.pos.line} col ${p.err.pos.offset}: ${p.err.toString()}`)
    }

    const globalScope = new ScopeMap()
    const aFunc: func[] = []

    const funcs: Map<string, FunctionDescription> = new Map()
    const globs: Map<string, GlobalObject> = new Map()

    p.ast.forEach(g => {
        const name = g.value.name.name
        if (globalScope.has(name)) {
            throw Error(`Another global or function is already called ${name}`)
        }
        
        switch (g.value.kind) {
        
            case ASTKinds.varDecl: 
                if (g.value.mutability !== "const") {
                    throw Error(`Global variable ${name} must be const`)
                }
                
                switch (g.value.value.root.kind) {
                    case ASTKinds.obj: 
                        if (g.value.value.root.fields.value.length > 0) {
                            break
                        }
                        if (g.value.value.methods.length > 0) {
                            break
                        }
                        globalScope.set(name, "global")
                        globs.set(name, {kind: "glob", name})
                        return
                    default: 
                        break
                }

                throw Error(`Global ${name} must be initialized as empty object`)
                
            case ASTKinds.func: 
                globalScope.set(name, "func")
                aFunc.push(g.value)
                break
            default: 
                const ne: never = g.value
        }
    })


    aFunc.forEach(f => funcs.set(f.name.name, to_descr(f, globalScope)))

    return  {
        globals: globs,
        funcs
    }


}