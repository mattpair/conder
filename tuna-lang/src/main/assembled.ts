import {FUNCTION_TRANSFORMER, FunctionDescription, Transformer ,Compiler} from 'conder_core'
import {Parser} from './parser'
import {semantify, Manifest} from './semantics'

export const TUNA_TO_MANIFEST = new Transformer<string, Manifest>(str => {
    const p = new Parser(str).parse()
    return semantify(p)
})