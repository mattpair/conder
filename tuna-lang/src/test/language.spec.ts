import { Parser } from '../main/parser';


describe("language", () => {

    function tunaTest(maybeSucceed: "succeed" | "fail", code: string): jest.ProvidesCallback {
        return (cb) => {
            if (maybeSucceed === "succeed") {
                const p = new Parser(code)
                const r = p.parse()
                expect(r).toMatchSnapshot()
            }
            cb()
        }
    }

    it("should allow a global object", tunaTest("succeed", `const obj = {}`))
        

    it("should allow many global objects",
        tunaTest("succeed", `
        const obj1 = {}
        const obj2 = {}
        `)
    )

    it("should allow empty public functions", 
    tunaTest("succeed", `public function doSomething() {}`))

    it("should allow a fixed number of args in functions",
    tunaTest("succeed", `public function argy(a, b, c) {

    }`)
    )

    it("should allow return statements within functions",
    tunaTest("succeed", `public function returny() {
        return
    }`))

    it("should allow setting of keys on a global object", 
    
        tunaTest("succeed",
        `
        const gg = {}
        public function fff(a) {
            gg.abc = a
            gg[a] = a
            gg['abc'] = a
        }
        `)
    )

    it("should allow getting of nested keys",
    
        tunaTest("succeed",
        `
        const gg = {}
        public function fff(a) {
            return gg[a].field
        }
        `)
    )

    it('should allow bools, numbers, and strings', tunaTest("succeed", `
    
    public function fff(a) {
        true
        false
        12
        -12.12
        'hello world'
        {}
    }
    
    `))

    it('can declare temp variables', tunaTest("succeed", `
    
    public function fff(a) {
        const b = true
        let c = false
        const d = a[b]
    }
    
    `))
})