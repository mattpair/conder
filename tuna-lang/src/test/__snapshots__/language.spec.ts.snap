// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`language can declare temp variables 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "args": Object {
          "kind": "args",
          "lastArg": Object {
            "kind": "name",
            "name": "a",
          },
          "leadingArgs": Array [],
        },
        "body": Object {
          "kind": "executable",
          "value": Array [
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "name",
                  "name": "const",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "assignment",
                "target": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "name",
                    "name": "b",
                  },
                },
                "value": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "bool",
                    "value": "true",
                  },
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "name",
                  "name": "let",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "assignment",
                "target": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "name",
                    "name": "c",
                  },
                },
                "value": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "bool",
                    "value": "false",
                  },
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "name",
                  "name": "const",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "assignment",
                "target": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "name",
                    "name": "d",
                  },
                },
                "value": Object {
                  "kind": "expression",
                  "methods": Array [
                    Object {
                      "kind": "method",
                      "method": Object {
                        "kind": "parameterIndex",
                        "value": Object {
                          "kind": "expression",
                          "methods": Array [],
                          "root": Object {
                            "kind": "name",
                            "name": "b",
                          },
                        },
                      },
                    },
                  ],
                  "root": Object {
                    "kind": "name",
                    "name": "a",
                  },
                },
              },
            },
          ],
        },
        "kind": "func",
        "name": Object {
          "kind": "name",
          "name": "fff",
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow a fixed number of args in functions 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "args": Object {
          "kind": "args",
          "lastArg": Object {
            "kind": "name",
            "name": "c",
          },
          "leadingArgs": Array [
            Object {
              "kind": "args_$0",
              "name": Object {
                "kind": "name",
                "name": "a",
              },
            },
            Object {
              "kind": "args_$0",
              "name": Object {
                "kind": "name",
                "name": "b",
              },
            },
          ],
        },
        "body": Object {
          "kind": "executable",
          "value": Array [],
        },
        "kind": "func",
        "name": Object {
          "kind": "name",
          "name": "argy",
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow a global object 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "kind": "varDecl",
        "mutability": "const",
        "name": Object {
          "kind": "name",
          "name": "obj",
        },
        "value": Object {
          "kind": "expression",
          "methods": Array [],
          "root": Object {
            "fields": Object {
              "kind": "fields",
              "value": Array [],
            },
            "kind": "obj",
          },
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow bools, numbers, and strings 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "args": Object {
          "kind": "args",
          "lastArg": Object {
            "kind": "name",
            "name": "a",
          },
          "leadingArgs": Array [],
        },
        "body": Object {
          "kind": "executable",
          "value": Array [
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "bool",
                  "value": "true",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "bool",
                  "value": "false",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "num",
                  "value": "12",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "num",
                  "value": "-12.12",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "kind": "str",
                  "value": "hello world",
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [],
                "root": Object {
                  "fields": Object {
                    "kind": "fields",
                    "value": Array [],
                  },
                  "kind": "obj",
                },
              },
            },
          ],
        },
        "kind": "func",
        "name": Object {
          "kind": "name",
          "name": "fff",
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow empty public functions 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "args": Object {
          "kind": "args",
          "lastArg": null,
          "leadingArgs": Array [],
        },
        "body": Object {
          "kind": "executable",
          "value": Array [],
        },
        "kind": "func",
        "name": Object {
          "kind": "name",
          "name": "doSomething",
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow getting of nested keys 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "kind": "varDecl",
        "mutability": "const",
        "name": Object {
          "kind": "name",
          "name": "gg",
        },
        "value": Object {
          "kind": "expression",
          "methods": Array [],
          "root": Object {
            "fields": Object {
              "kind": "fields",
              "value": Array [],
            },
            "kind": "obj",
          },
        },
      },
    },
    Object {
      "kind": "globals_$0",
      "value": Object {
        "args": Object {
          "kind": "args",
          "lastArg": Object {
            "kind": "name",
            "name": "a",
          },
          "leadingArgs": Array [],
        },
        "body": Object {
          "kind": "executable",
          "value": Array [
            Object {
              "kind": "executable_$0",
              "value": "return",
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "expression",
                "methods": Array [
                  Object {
                    "kind": "method",
                    "method": Object {
                      "kind": "parameterIndex",
                      "value": Object {
                        "kind": "expression",
                        "methods": Array [],
                        "root": Object {
                          "kind": "name",
                          "name": "a",
                        },
                      },
                    },
                  },
                  Object {
                    "kind": "method",
                    "method": Object {
                      "kind": "literalIndex",
                      "value": Object {
                        "kind": "name",
                        "name": "field",
                      },
                    },
                  },
                ],
                "root": Object {
                  "kind": "name",
                  "name": "gg",
                },
              },
            },
          ],
        },
        "kind": "func",
        "name": Object {
          "kind": "name",
          "name": "fff",
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow many global objects 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "kind": "varDecl",
        "mutability": "const",
        "name": Object {
          "kind": "name",
          "name": "obj1",
        },
        "value": Object {
          "kind": "expression",
          "methods": Array [],
          "root": Object {
            "fields": Object {
              "kind": "fields",
              "value": Array [],
            },
            "kind": "obj",
          },
        },
      },
    },
    Object {
      "kind": "globals_$0",
      "value": Object {
        "kind": "varDecl",
        "mutability": "const",
        "name": Object {
          "kind": "name",
          "name": "obj2",
        },
        "value": Object {
          "kind": "expression",
          "methods": Array [],
          "root": Object {
            "fields": Object {
              "kind": "fields",
              "value": Array [],
            },
            "kind": "obj",
          },
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow return statements within functions 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "args": Object {
          "kind": "args",
          "lastArg": null,
          "leadingArgs": Array [],
        },
        "body": Object {
          "kind": "executable",
          "value": Array [
            Object {
              "kind": "executable_$0",
              "value": "return",
            },
          ],
        },
        "kind": "func",
        "name": Object {
          "kind": "name",
          "name": "returny",
        },
      },
    },
  ],
  "err": null,
}
`;

exports[`language should allow setting of keys on a global object 1`] = `
ParseResult {
  "ast": Array [
    Object {
      "kind": "globals_$0",
      "value": Object {
        "kind": "varDecl",
        "mutability": "const",
        "name": Object {
          "kind": "name",
          "name": "gg",
        },
        "value": Object {
          "kind": "expression",
          "methods": Array [],
          "root": Object {
            "fields": Object {
              "kind": "fields",
              "value": Array [],
            },
            "kind": "obj",
          },
        },
      },
    },
    Object {
      "kind": "globals_$0",
      "value": Object {
        "args": Object {
          "kind": "args",
          "lastArg": Object {
            "kind": "name",
            "name": "a",
          },
          "leadingArgs": Array [],
        },
        "body": Object {
          "kind": "executable",
          "value": Array [
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "assignment",
                "target": Object {
                  "kind": "expression",
                  "methods": Array [
                    Object {
                      "kind": "method",
                      "method": Object {
                        "kind": "literalIndex",
                        "value": Object {
                          "kind": "name",
                          "name": "abc",
                        },
                      },
                    },
                  ],
                  "root": Object {
                    "kind": "name",
                    "name": "gg",
                  },
                },
                "value": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "name",
                    "name": "a",
                  },
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "assignment",
                "target": Object {
                  "kind": "expression",
                  "methods": Array [
                    Object {
                      "kind": "method",
                      "method": Object {
                        "kind": "parameterIndex",
                        "value": Object {
                          "kind": "expression",
                          "methods": Array [],
                          "root": Object {
                            "kind": "name",
                            "name": "a",
                          },
                        },
                      },
                    },
                  ],
                  "root": Object {
                    "kind": "name",
                    "name": "gg",
                  },
                },
                "value": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "name",
                    "name": "a",
                  },
                },
              },
            },
            Object {
              "kind": "executable_$0",
              "value": Object {
                "kind": "assignment",
                "target": Object {
                  "kind": "expression",
                  "methods": Array [
                    Object {
                      "kind": "method",
                      "method": Object {
                        "kind": "parameterIndex",
                        "value": Object {
                          "kind": "expression",
                          "methods": Array [],
                          "root": Object {
                            "kind": "str",
                            "value": "abc",
                          },
                        },
                      },
                    },
                  ],
                  "root": Object {
                    "kind": "name",
                    "name": "gg",
                  },
                },
                "value": Object {
                  "kind": "expression",
                  "methods": Array [],
                  "root": Object {
                    "kind": "name",
                    "name": "a",
                  },
                },
              },
            },
          ],
        },
        "kind": "func",
        "name": Object {
          "kind": "name",
          "name": "fff",
        },
      },
    },
  ],
  "err": null,
}
`;
